### ISSUE TITLE
AbortController, AbortSignal, and garbage collection

### ISSUE THREAD

**Issue by annevk** [whatwg/dom#920]
@smaug---- made me realize we could probably be a little bit more specific about what keeps what alive.

Thoughts:

* AbortController weakly references AbortSignal. The controller getting GC'd is fine, since it doesn't learn about results.
* AbortSignal cannot be collected when it has a controller and event listeners/abort algorithms. I think that means it needs a weak reference to its controller.

cc @jakearchibald @domenic

**Comment by jakearchibald** [comment]
The controller already has a strong reference to the signal, no? Via `controller.signal`. I guess that value could be overridden, but do we generally try to cater for cases like that?

**Comment by annevk** [comment]
Sorry, yeah, the controller keeps the signal alive, but not the other way around. I'll fix OP.

**Comment by jakearchibald** [comment]
Abort algorithms need to stick around while the controller is in reference and the signal is not aborted.

Abort events listeners need to stick around while the signal is in reference, or the controller is in reference and the signal is not aborted. But some of that seems redundant given the controller has a strong reference to the signal.

**Comment by annevk** [comment]
Oh yeah, I think the basic scenarios are fine already, indeed. My bad.

I guess the one thing that complicates things somewhat might be signals that follow other signals.

**Comment by MattiasBuelens** [comment]
> I guess the one thing that complicates things somewhat might be signals that follow other signals. Edit: would that be as simple as keeping it alive from the signal that it is following? At least as long as the aborted flag isn't set.

Again, isn't that already the case by virtue of the *followingSignal* attaching abort steps to its *parentSignal*, and those steps necessarily keeping the *followingSignal* alive?

**Comment by MattiasBuelens** [comment]
On a slightly unrelated note: what **is** missing here is if the *followingSignal* is aborted separately (i.e. through its own abort controller, not as a result of the *parentSignal* becoming aborted). In that case, the *followingSignal* should remove the abort steps it had attached to its *parentSignal*, so that the *parentSignal* does not keep the *followingSignal* alive unnecessarily.

I don't think this shouldn't be too difficult to fix in the specification. I'll give it a try.

**Comment by annevk** [comment]
Hmm yeah. I'm not sure you can ever be a _followingSignal_ with your own controller. Fetch uses this mechanism and always creates an "unowned" signal for this (as part of creating a new `Request` object).

**Comment by MattiasBuelens** [comment]
So either a specification calls "signal abort" on a signal, or it makes it "follow" another signal, but it can't ever do both on the same signal? I guess that makes sense, it's just not explicitly documented as a restriction. Not sure if we need to. 🤷

I've been using an `AbortController`/`AbortSignal` implementation in my own code, and I sometimes find it useful being able to do both on the same signal. For example, I'd have a "long-living" abort signal for the lifetime of an entire component, and then create "short-living" signals for tasks spawned by that component. These task signals can be aborted separately, but they also follow the long-living signal so all tasks are automatically aborted when the component is destroyed. But then again, this might be a very specific use case that isn't very useful for the whole of the web platform. 😅

Anyway, sorry for the sidetrack. So yeah, I *think* GC is already pretty okay for signals following other signals.

**Comment by jakearchibald** [comment]
> But then again, this might be a very specific use case that isn't very useful for the whole of the web platform.

I had a use-case like this a couple of days ago fwiw. It was on squoosh.app, where there's a single controller that does overall things like decoding the input image, but then there are two other controllers for processing and encoding each side. Changing the encoding settings on one side only aborts the operations on that side, but dropping in a new image aborts main image decoding, and processing of both sides.

**Comment by annevk** [comment]
I guess we're really discussing a new issue now, but that's fine with me.

1. Following is not exposed, but we should perhaps document the limitation in case a specification tries to use it for something it doesn't work. I'd rather do that than make it work as it would be redundant at the moment.
2. Does following need to be exposed? It seems it should be doable to implement the scenarios you sketch in userland?

**Comment by MattiasBuelens** [comment]
Indeed, it's easy enough to do in userland, and I'm totally fine with keeping it there:
```javascript
function followSignal(followingAbortController, parentSignal) {
  // 1. If followingSignal’s aborted flag is set, then return.
  if (followingAbortController.aborted) {
    return;
  }
  // 2. If parentSignal’s aborted flag is set, then signal abort on followingSignal.
  if (parentSignal.aborted) {
    followingAbortController.abort();
    return;
  }
  // 3. Otherwise, add the following abort steps to parentSignal:
  const onAbort = () => {
    // 3.1. Signal abort on followingSignal.
    followingAbortController.abort();
  };
  parentSignal.addEventListener('abort', onAbort, { once: true });
  // Non-standard: if followingSignal becomes aborted for some other reason,
  // remove the abort steps from the parentSignal.
  followingAbortController.signal.addEventListener('abort', () => {
    parentSignal.removeEventListener('abort', onAbort);
  });
}
```
That said, I wouldn't mind having this exposed as `AbortController.prototype.follow(signal)` if there's enough interest. 🙂

**Comment by noseratio** [comment]
> That said, I wouldn't mind having this exposed as `AbortController.prototype.follow(signal)` if there's enough interest. 🙂

Yes please! :) Coming from .NET, I use hierarchical cancellation all the time, it's very handy. I currently do it with [`CancellationTokenSource(linkedTokens?)`](https://github.com/rbuckton/prex/blob/master/docs/cancellation.md#new-cancellationtokensourcelinkedtokens) from @rbuckton's [Prex](https://github.com/rbuckton/prex) for that.

**Comment by MattiasBuelens** [comment]
I just realized that if we have #911, this could be simplified(?) to:
```javascript
  // 3. Otherwise, add the following abort steps to parentSignal:
  parentSignal.addEventListener('abort',  () => {
    // 3.1. Signal abort on followingSignal.
    followingAbortController.abort();
  }, {
    once: true,
    signal: followingAbortController.signal
  });
```
It's abort signals all the way down! 😆

**Comment by benjamingr** [comment]
Hey, a few of us have been discussing this a lot and we think there needs to be a way for signals to "follow" other signals. This is because making linked signals is _very_ hard at the moment.

This is [impossible to do in userland](https://github.com/nodejs/node/issues/35990#issuecomment-723154000) though I'll let Brian leave a more detailed comment.

Node is very interested in this API and I believe it is very hard to use AbortSignal for large web apps with "linked signals" without it. The use case we've seen over and over is something like [this one](https://github.com/benjamingr/cancellation-use-cases/blob/main/use-cases/01-server-process-request.md) - I've actually seen this in browser environments and converted it to a Node case for our meeting :]

Regarding the color of the bike shed: Is there any reason this is `AbortController.prototype.follow(signal)` and not `new AbortController(...signals)`?

**Comment by benjamingr** [comment]
@MattiasBuelens hey :] Note your implementation of `followSignal` does not work in case the signal is _never_ aborted. It works for the "operation was aborted" case but not for the "everything went fine" case.

The handler would have to be weak so that if the operation completes successfully the parent controller's signal does not keep a reference to the child controller. This is impossible in userland because `addEventListener`'s API takes a function.

**Comment by bterlson** [comment]
@MattiasBuelens fwiw I don't believe this is particularly easy to do in userland without leaking signals. Consider an application which has a top-level signal that lasts basically forever (e.g. its job is to tell the application its about to go down so cancel any ongoing work and clean up). Every signal in the application follows that signal to handle more granular events like user input or timeouts. With the implementation above, assuming the happy path of no cancellation occurs, the parent signal keeps all its children signals alive forever.

Follow signals are critical so we should definitely figure this out, and happy to see more interest along these lines! I have concerns about the proposed API considering linking multiple signals together (e.g. a cancel button plus timeout plus sigint listener) can be fairly common.

**Comment by wanderview** [comment]
Sorry if the conversation has moved on, but:

> Hmm yeah. I'm not sure you can ever be a followingSignal with your own controller.

I implemented something like this in the browser as well.  For cache.addAll() I needed to abort outstanding requests if one of them failed.  The spec just uses verbiage like "terminate all ongoing fetches with the aborted flag set", but in practice the most direct way to do it was with an internally created AbortSignal and AbortController.  And since I still needed to honor individual request.signal I chained them together with follow.

**Comment by benjamingr** [comment]
@wanderview just to add to what you're saying, when you [follow](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/cache_storage/cache.cc;l=1089;drc=ae1a631ceaf488f5e9cb55755de32b6a505b4b1e) you basically use the capability we're asking for that isn't exposed to userland.

In userland, you can't chain signals in a way that doesn't require `addEventListener` which means the parent signal always holds a reference to the child controller (which can live for much longer) - this is the memory leak we've run into :]

**Comment by MattiasBuelens** [comment]
Thanks for chiming in! Indeed, from my experience it's very easy to leak abort event listeners in user code.

In my custom `AbortController` implementation, I have an additional `destroy()` method on `AbortController`. This first fires a `'destroy'` event on its signal, and then removes all `'abort'` and `'destroy'` listeners that were attached on that signal. (I know, this isn't possible with a regular `EventTarget`, so I'm also rolling my own implementation there. 😛).

For my custom `AbortController.follow()` implementation, I add a `'destroy'` listener to **both** the parent signal and the following signal, which removes all listeners that were added to either signal. This ensures that all the "links" between the two signals are removed, so they no longer keep each other alive.

It's messy and tedious, since I need to carefully add a bunch of `destroy()` calls throughout my code whenever I'm "done" with a certain signal. For example, this pattern pops up several times:
```javascript
const tempController = new MyAbortController();
tempController.follow(signal);
try {
  await doSomething(tempController.signal);
} finally {
  // At this point, either the controller became aborted which caused `doSomething` to reject,
  // or `doSomething` resolved/rejected normally. So we can clean up our temporary abort controller.
  tempController.destroy();
}
```

I'd be *very* interested in ways to get rid of this mess. 😁

**Comment by bterlson** [comment]
@MattiasBuelens careful... `destroy` on a controller doesn't necessarily mean its associated signal and its listeners can be disposed. I explained this some [here](https://github.com/nodejs/node/issues/35990#issuecomment-723218855).

**Comment by domenic** [comment]
A few points:

1. I agree this would be a good thing to add, if only because so many users are asking for it. This holds regardless of my next two points.

1. The "follow" operation has always seemed unintuitive to me, compared with a "race" or "create linked token source" style API ([described here in the context of the defunct cancelable promises proposal](https://github.com/tc39/proposal-cancelable-promises/blob/0e769fda8e16bff0feffe964fddc43dcd86668ba/Cancel%20Tokens.md#combining-cancel-tokens)). But, I don't trust my inutition here. Would someone with more experience be able to discuss the differences between these, and argue for one or the other?

1. Even if we add this high-level combinator, it'd be ideal if it could be expressed in userland code. There are a few potential missing primitives here, and I believe any one of them would suffice:
   - Weak event listeners. @syg and I discussed this in some private chat some months ago. You can get close to emulating these today with `WeakRef`, but you'll leak a bit of memory for the wrapper thunk, which looks like `e => weakRef.get().?(e)`.
   
   - Some way of clearing all event listeners for the signal's `"abort"` event after the controller's `abort()` is called. E.g. https://github.com/whatwg/dom/issues/412 would allow this. If we exposed this then I'd suggest `abort()` auto-clears all listeners. (Which would be a backward-incompatible change, but hopefully a safe one.)
   
   - Probably some other ideas.

1. I don't fully understand the relationship between the issues discussed here, and how fetch() and other platform features ignore the event listener mechanisms in favor of their own internal "abort steps". But, in the past I've advocated against this magic; see https://github.com/whatwg/dom/issues/493 and the somewhat-related https://github.com/whatwg/dom/issues/878 .

**Comment by benjamingr** [comment]
Re: API - I think the simplest API would probably be a constructor argument:

```js
const childController = new AbortController(parentSignal); // creates a linked AbortController
```

 I'd love to see a use case where adding the reference to the child controller at construction doesn't work.

Re: weak event listeners and WeakRefs - I want to agree and echo there is no way to not "leak the thunk" - I get Node servers are not really a priority for WHATWG but I assume people with long running PWAs can likely run into this issue even if they implement the WeakRef solution you propose in userland (note you also need to keep a strong reference from the child to the parent in that case). Weak event listeners would address this as well though they feel like a much more powerful tool.

**Comment by rbuckton** [comment]
> 2. The "follow" operation has always seemed unintuitive to me, compared with a "race" or "create linked token source" style API ([described here in the context of the defunct cancelable promises proposal](https://github.com/tc39/proposal-cancelable-promises/blob/0e769fda8e16bff0feffe964fddc43dcd86668ba/Cancel%20Tokens.md#combining-cancel-tokens)). But, I don't trust my inutition here. Would someone with more experience be able to discuss the differences between these, and argue for one or the other?

`race` is the approach I went with in https://esfx.js.org/esfx/api/async-canceltoken/canceltoken.html#_esfx_async_canceltoken_CancelToken_race_member_1_ (which is the successor to `prex` and based on my [Cancellation Protocol](https://github.com/tc39/proposal-cancellation/issues/22) proposal).

**Comment by bterlson** [comment]
@domenic regarding 3.b, a big part of the problem is listeners sticking around in non-abort situations, so by itself clearing all listeners on controller abort doesn't solve the problem I think?

**Comment by MattiasBuelens** [comment]
> @MattiasBuelens `destroy` on a controller doesn't necessarily mean its associated signal and its listeners can be disposed. I explained this some [here](https://github.com/nodejs/node/issues/35990#issuecomment-723218855).

I see. For my use cases, a child signal following a parent signal never needs to "outlive" its parent signal, so it's fine for me to destroy the parent signal (and therefore all links with child signals) once I'm done with it. But I understand that this may not apply to all use cases, and a more general solution would be desirable.

**Comment by benjamingr** [comment]
@domenic 
> I tried to link to a couple examples of that, where you want to create a new AbortSignal from two other AbortSignals. It's not obvious to me how to do that from a single follow like you describe.

I think `new AbortController(...signals)` is actually very simple to `CancelToken.race` API in that the construction is at the creation. For example the first example is:

```js
const ac1 = new AbortController();
const ac2 = new AbortController();

const combined = new AbortController(ac1.signal, ac2.signal);
```

Though in this case, since there is a hierarchy you don't actually usually create 3 sources but two.

I think doing this on the signal and not controller can work though having users write:

```js
const parent = new AbortController();
const child = new AbortController();
// ...
action(AbortSignal.race(parent.signal, child.signal));

const parent = new AbortController();
// child controller
const child = new AbortController(parent.signal);
// ...
action(child.signal);
```

Is a bit more code and doesn't establish hierarchy between the parent/child controllers (which may be desireable?).

**Comment by benjamingr** [comment]
What is a good way to make progress here?
Would it help if I work on the spec?
Do implementors also agree that this is something that this change is good?

What about a meeting to discuss API alternatives?

**Comment by annevk** [comment]
I wonder if we should flush out the use cases a bit more to ensure we are all on the same page and have something to evaluate a proposal against. Precedent in other ecosystems would also be great to see. Is there anything besides .NET?

(This definitely sounds like the kind of thing implementers would be happy to add though, nobody likes memory leaks.)

**Comment by benjamingr** [comment]
@annevk yes, I actually have a list at https://github.com/benjamingr/cancellation-use-cases - I will adapt it to parent/child relationships.

**Comment by domenic** [comment]
This feels like it's the prototypical example of where an [explainer](https://w3ctag.github.io/explainers) would make sense. (Also, this is probably the sort of thing where if we were to ship it in Chrome, we'd require an explainer; it's not as much of a no-brainer small addition as #911, and so a background document is necessary.)

Logistically, if it helps, you can consider an explainer similar to the README of a TC39 proposal. Also, it doesn't matter where the explainer is posted; you could put it in your cancellation-use-cases repo as a separate document.

To be clear, I think a lot of people on this thread are on board with the general idea. But the problem and solution space is complicated enough that it would benefit from someone taking the time to collate information like:

- Example use cases, including simple code samples so that people get the idea, plus a sketch of more complicated real-world situations

- Comparison with other ecosystems

- Discussing how people solve this today, and what the drawbacks are.

- Comparison between different APIs, e.g. `new AbortController(...signals)` vs. `new AbortSignal(...signals)` vs. `AbortSignal.race([...signals])` vs. `abortSignal.follow(otherSignal)`. Eventually the explainer should pick one as favored and make the case for it, relegating the others to an "Alternatives considered" section. But maybe you want to start in an exploratory state that puts them all on equal footing and compares them.

- Mentioning other alternative ways to solve this problem, e.g. my ideas (3) in https://github.com/whatwg/dom/issues/920#issuecomment-726961899 .

A lot of this information is scattered throughout the conversation in these issue threads, but collating it into one place will be helpful for outsiders (including, e.g., persuading various implementation teams who might not be following this thread).

**Comment by noseratio** [comment]
> I wonder if we should flush out the use cases

@annevk, here's one I've been recently experimenting with: `AbortablePromise`. It creates a child `AbortController` to help with the proper garbage collection of event listeners, because the child `AbortController`/`AbortSignal` will go out of scope when the promise is resolved. 

In a nutshell ([full gist](https://gist.github.com/noseratio/b7de3c7383ae3660904d650cd418c2dd)):

```JavaScript
function delay(ms, outerSignal) {
  return new AbortablePromise(outerSignal, ({resolve, signal}) => {
    const id = setTimeout(resolve, ms);
    signal.addEventListener('abort', () => clearTimeout(id));
  });  
} 

const controller = new AbortController();
setTimeout(() => controller.abort(), 1000);
await delay(2000, controller.signal);
```

I haven't checked yet if everything gets GC'ed as expected, but it certainly would not without a child controller. 

This experiment was inspired by @jakearchibald's ["Event listeners and garbage collection"](https://jakearchibald.com/2020/events-and-gc/) - thanks Jake!

**Comment by benjamingr** [comment]
@domenic thanks that sounds very reasonable. I'll work on that (probably next weekend)

**Comment by benlesh** [comment]
Reading through some of the older comments in this thread, I just wanted to point out that every single use case I have requires the following signal to have its own controller so it can be aborted separately, which removes the abort logic that was added to the parent.

**Comment by bart7951** [comment]
> Re: API - myślę, że najprostsze API byłoby prawdopodobnie argumentem konstruktora:
> 
> ```js
> const  childController  =  new  AbortController ( parentSignal ) ;  // tworzy połączony AbortController
> ```
> 
> Chciałbym zobaczyć przypadek użycia, w którym dodanie odwołania do kontrolera podrzędnego podczas tworzenia nie działa.
> 
> Re: słabe nasłuchiwacze zdarzeń i WeakRefs - chcę się zgodzić i powtórzyć, że nie ma sposobu, aby nie „wyciekać” - rozumiem, że serwery Node nie są tak naprawdę priorytetem WHATWG, ale zakładam, że ludzie z długo działającymi PWA mogą prawdopodobnie natknąć się na to problem, nawet jeśli implementują rozwiązanie WeakRef, które proponujesz w przestrzeni użytkownika (pamiętaj, że w tym przypadku musisz również zachować silne odniesienie od dziecka do rodzica). Słabi słuchacze wydarzeń również by się tym zajęli, chociaż czują się znacznie potężniejszym narzędziem.

hallo komm bitte bei mir .. sex kamera ich mochte heute ,, kann sein skeyp      LeolDe Kik  das nick

**Comment by noseratio** [comment]
> The handler would have to be weak so that if the operation completes successfully the parent controller's signal does not keep a reference to the child controller. This is impossible in userland because `addEventListener`'s API takes a function.

@benjamingr I've been playing with this idea using `WeakRef` and `FinalizationRegistry`, and I think I may have come up with [something that works](https://github.com/nodejs/node/discussions/36467), although it isn't trivial to test.

**Comment by benjamingr** [comment]
@noseratio I made a PR (to Node) for adding weak listeners (internally, and as part of working on the explainer above which I am just getting to due to a surprise vacation my wife ordered). https://github.com/nodejs/node/pull/36607

**Comment by noseratio** [comment]
> I made a PR (to Node) for adding weak listeners

@benjamingr this is awesome, thank you! I believe this is going to be really useful beyond `AbortController`, too. I used the [weak event pattern](https://docs.microsoft.com/en-us/dotnet/desktop/wpf/advanced/weak-event-patterns?view=netframeworkdesktop-4.8) quite a few times in .NET. 

Once your PR is merged in, what would be best way to detect this feature? I have this [`createWeakEventListener`](https://gist.github.com/noseratio/f037cd08041a72e31e3899747cb85462#file-linkedabortcontroller-mjs-L15) that works in browsers, for Node I'd like to update it to use your implementation where available.

**Comment by benjamingr** [comment]
@noseratio at the moment I want to not add any APIs and work [understanding use cases better](https://github.com/benjamingr/cancellation-use-cases/issues/9). So unfortunately `kWeakHandler` is entirely internal to Node.js core at the moment.

I feel strongly about doing this as part of a web standard effort rather than Node providing its own API :] 

I would love contributions in [the use cases](https://github.com/benjamingr/cancellation-use-cases/issues/9). I intend to start writing things down and I'll probably invite people to meetings a few times :]

**Comment by benjamingr** [comment]
Hey, I just wanted to let you know that in the past month I've spent a bunch of time on this - I've rewritten the Node weak event listeners PR around... 3 times now.

I think this issue is tricker than one might expect. I am also considering several other alternatives (like a utility `aborted` method that does not retain the listener to allow chaining without leaks).

I'm just not very good at this yet so this is taking me time :] I just wanted to update that this is still very much something I'm spending time on and that any help and discussion would be appreciated.

**Comment by shaseley** [comment]
Hi all, we're (Chrome) interested in tackling this, and I've put together an explainer for [`AbortSignal.any(signals)`](https://github.com/shaseley/abort-signal-any/blob/main/README.md) as a solution for combining `AbortSignal`s. Feedback would be much appreciated, either as issues on that repo or here. Thanks!

**Comment by annevk** [comment]
Thanks @shaseley, I opened an issue there, mainly to suggest an additional requirement for this feature that it either uses or obsoletes the current "follow" primitive in the specification.

@benlesh I wanted to ask if you could elaborate on what you wrote above:

> [E]very single use case I have requires the following signal to have its own controller so it can be aborted separately, which removes the abort logic that was added to the parent.

In particular I'm not sure I understand what you mean by the very last part of that sentence.

**Comment by jasnell** [comment]
`AbortSignal.any(signal)` seems reasonable. I can't really imagine many cases where I would pass more than 2 signals into it but the array input is acceptable. Implementation should be straightforward and fairly easy to optimize. The most compelling use case I see is creating a single that either times out or is explicitly triggered.

**Comment by vitalets** [comment]
Shouldn't we consider `AbortController.unfollow(signal)`?

**Comment by keithamus** [comment]
What do you expect `AbortController.unfollow(signal)` to do?

**Comment by vitalets** [comment]
> What do you expect `AbortController.unfollow(signal)` to do?

Stops following parent signal: child signal not aborted when parent is aborted.

**Comment by keithamus** [comment]
What would be the use case for unfollowing a signal?

**Comment by vitalets** [comment]
> What would be the use case for unfollowing a signal?

Imagine 2 http requests that should be aborted by timeout. First request successfully passed, and in that case we want to wait second request anyway to have app in some desired state. if second request's abort controller is following the first one, it will be aborted by timeout.

I just want to highlight that is followed AbortControllers are passed as constructor arguments, there is no way to unfollow them.
